File: taurus-protect-sdk-typescript/src/models/wallet.ts
/**
 * Wallet domain models for Taurus-PROTECT SDK.
 */

import type { Currency } from './currency';

/**
 * Wallet status enum.
 */
export enum WalletStatus {
  UNKNOWN = 'UNKNOWN',
  ACTIVE = 'ACTIVE',
  PENDING = 'PENDING',
  DISABLED = 'DISABLED',
}

/**
 * Wallet type enum.
 */
export enum WalletType {
  UNKNOWN = 'UNKNOWN',
  STANDARD = 'STANDARD',
  AIRGAP = 'AIRGAP',
  STAKING = 'STAKING',
}

/**
 * Represents wallet balance information.
 */
export interface WalletBalance {
  /** Total confirmed balance in the smallest currency unit (e.g., WEI for ETH) */
  readonly totalConfirmed: string | undefined;
  /** Total balance including unconfirmed transactions */
  readonly totalUnconfirmed: string | undefined;
  /** Available confirmed balance ready to be spent */
  readonly availableConfirmed: string | undefined;
  /** Available balance including unconfirmed transactions */
  readonly availableUnconfirmed: string | undefined;
  /** Confirmed reserved balance held for pending transactions */
  readonly reservedConfirmed: string | undefined;
  /** Reserved unconfirmed balance not yet validated */
  readonly reservedUnconfirmed: string | undefined;
}

/**
 * Represents a wallet attribute (key-value metadata).
 */
export interface WalletAttribute {
  /** Unique attribute identifier */
  readonly id: string | undefined;
  /** Attribute key */
  readonly key: string | undefined;
  /** Attribute value */
  readonly value: string | undefined;
  /** Content type of the attribute */
  readonly contentType: string | undefined;
  /** Owner of the attribute */
  readonly owner: string | undefined;
  /** Attribute type */
  readonly type: string | undefined;
  /** Attribute subtype */
  readonly subtype: string | undefined;
  /** Whether this attribute is a file */
  readonly isFile: boolean | undefined;
}

/**
 * Represents a cryptocurrency wallet in Taurus-PROTECT.
 */
export interface Wallet {
  /** Unique identifier */
  readonly id: string;
  /** Human-readable name */
  readonly name: string;
  /** Current status */
  readonly status: WalletStatus;
  /** Wallet type */
  readonly type: WalletType;
  /** Blockchain identifier (e.g., "ETH", "BTC") */
  readonly blockchain: string | undefined;
  /** Network identifier (e.g., "mainnet", "testnet") */
  readonly network: string | undefined;
  /** Currency identifier or symbol */
  readonly currency: string | undefined;
  /** Whether this wallet is disabled */
  readonly disabled: boolean;
  /** Whether this is an omnibus wallet */
  readonly isOmnibus: boolean;
  /** Wallet balance information */
  readonly balance: WalletBalance | undefined;
  /** Creation timestamp */
  readonly createdAt: Date | undefined;
  /** Last update timestamp */
  readonly updatedAt: Date | undefined;
  /** Optional description/comment */
  readonly comment: string | undefined;
  /** Customer identifier for omnibus wallets */
  readonly customerId: string | undefined;
  /** Number of addresses in the wallet */
  readonly addressesCount: number | undefined;
  /** Wallet attributes (key-value metadata) */
  readonly attributes: WalletAttribute[];
  /** Visibility group ID */
  readonly visibilityGroupId: string | undefined;
  /** External wallet identifier */
  readonly externalWalletId: string | undefined;
  /** HD wallet account derivation path */
  readonly accountPath: string | undefined;
  /** Detailed currency information */
  readonly currencyInfo: Currency | undefined;
  /** Associated tags (for backward compatibility) */
  readonly tags: string[];
}

/**
 * Request parameters for creating a new wallet.
 */
export interface CreateWalletRequest {
  /**
   * Wallet name (must be unique per blockchain/network).
   * Required.
   */
  readonly name: string;
  /**
   * Blockchain identifier (e.g., "ETH", "BTC").
   * Required if currency is not provided.
   */
  readonly blockchain?: string;
  /**
   * Network identifier (e.g., "mainnet", "testnet").
   * Required if blockchain is enabled on multiple networks.
   */
  readonly network?: string;
  /**
   * Currency ID or symbol. Required if blockchain is not provided.
   * Must be the native currency of a blockchain.
   */
  readonly currency?: string;
  /**
   * Whether this is an omnibus wallet (single owner).
   * When true, addresses can be used as transaction sources.
   */
  readonly isOmnibus?: boolean;
  /**
   * Optional comment or description (max 254 characters).
   */
  readonly comment?: string;
  /**
   * Customer identifier (max 254 characters).
   * Applied to all addresses when isOmnibus is true.
   */
  readonly customerId?: string;
  /**
   * Visibility group UUID.
   * Restricts wallet visibility and management to users in this group.
   */
  readonly visibilityGroupId?: string;
  /**
   * External wallet identifier.
   */
  readonly externalWalletId?: string;
}

/**
 * Options for listing wallets.
 */
export interface ListWalletsOptions {
  /** Maximum number of wallets to return (default: 50) */
  readonly limit?: number;
  /** Number of wallets to skip for pagination (default: 0) */
  readonly offset?: number;
  /** Filter by currency ID or symbol */
  readonly currency?: string;
  /** Filter by currencies (array of IDs or symbols) */
  readonly currencies?: string[];
  /** Search query string */
  readonly query?: string;
  /** Filter by wallet name (case-insensitive partial match) */
  readonly name?: string;
  /** Sort order: "ASC" or "DESC" */
  readonly sortOrder?: string;
  /** Exclude wallets disabled by business rules */
  readonly excludeDisabled?: boolean;
  /** Filter by tag IDs (OR combination) */
  readonly tagIds?: string[];
  /** Only include wallets with positive balance */
  readonly onlyPositiveBalance?: boolean;
  /** Filter by blockchain */
  readonly blockchain?: string;
  /** Filter by network */
  readonly network?: string;
  /** Filter by specific wallet IDs (max 100) */
  readonly ids?: string[];
}

/**
 * Options for getting wallet balance history.
 */
export interface GetBalanceHistoryOptions {
  /** Interval in hours for balance snapshots */
  readonly intervalHours: number;
}

/**
 * A point in the wallet balance history.
 */
export interface BalanceHistoryPoint {
  /** Timestamp of this balance snapshot */
  readonly pointDate: Date | undefined;
  /** Balance at this point in time */
  readonly balance: WalletBalance | undefined;
}

File: taurus-protect-sdk-python/taurus_protect/models/wallet.py
"""Wallet models for Taurus-PROTECT SDK."""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field

from taurus_protect.models.balance import Balance
from taurus_protect.models.currency import Currency


class WalletAttribute(BaseModel):
    """Custom attribute on a wallet."""

    id: str = Field(description="Attribute identifier")
    key: str = Field(description="Attribute name")
    value: str = Field(description="Attribute value")
    content_type: Optional[str] = Field(default=None, description="Content type")
    owner: Optional[str] = Field(default=None, description="Attribute owner")
    type: Optional[str] = Field(default=None, description="Attribute type")
    subtype: Optional[str] = Field(default=None, description="Attribute subtype")
    is_file: bool = Field(default=False, description="Whether this is a file attribute")

    model_config = {"frozen": True}


class Wallet(BaseModel):
    """
    Cryptocurrency wallet.

    A wallet is a container for blockchain addresses of a specific currency.

    Attributes:
        id: Unique wallet identifier.
        name: Human-readable wallet name.
        currency: Currency symbol (e.g., "BTC", "ETH").
        blockchain: Blockchain name.
        network: Network name (e.g., "mainnet", "testnet").
        balance: Current wallet balance.
        is_omnibus: Whether this is an omnibus (pooled) wallet.
        disabled: Whether the wallet is disabled.
        comment: Optional description.
        customer_id: Optional customer identifier.
        external_wallet_id: Optional external identifier.
        visibility_group_id: Optional visibility group.
        account_path: HD wallet account derivation path.
        addresses_count: Number of addresses in the wallet.
        created_at: When the wallet was created.
        updated_at: When the wallet was last updated.
        attributes: Custom key-value attributes.
        currency_info: Detailed currency information.
    """

    id: str = Field(description="Unique wallet identifier")
    name: str = Field(description="Human-readable wallet name")
    currency: str = Field(description="Currency symbol")
    blockchain: str = Field(default="", description="Blockchain name")
    network: str = Field(default="", description="Network name")
    balance: Optional[Balance] = Field(default=None, description="Current wallet balance")
    is_omnibus: bool = Field(default=False, description="Whether this is an omnibus wallet")
    disabled: bool = Field(default=False, description="Whether the wallet is disabled")
    comment: Optional[str] = Field(default=None, description="Optional description")
    customer_id: Optional[str] = Field(default=None, description="Optional customer identifier")
    external_wallet_id: Optional[str] = Field(default=None, description="Optional external ID")
    visibility_group_id: Optional[str] = Field(default=None, description="Visibility group ID")
    account_path: Optional[str] = Field(default=None, description="HD account derivation path")
    addresses_count: int = Field(default=0, description="Number of addresses")
    created_at: Optional[datetime] = Field(default=None, description="Creation timestamp")
    updated_at: Optional[datetime] = Field(default=None, description="Last update timestamp")
    attributes: List[WalletAttribute] = Field(default_factory=list, description="Custom attributes")
    currency_info: Optional[Currency] = Field(default=None, description="Detailed currency information")

    model_config = {"frozen": True}


class CreateWalletRequest(BaseModel):
    """
    Request to create a new wallet.

    Example:
        >>> request = CreateWalletRequest(
        ...     blockchain="ETH",
        ...     network="mainnet",
        ...     name="Trading Wallet",
        ...     is_omnibus=True,
        ...     comment="Primary trading account",
        ... )
    """

    blockchain: str = Field(description="Blockchain type (e.g., 'ETH', 'BTC', 'SOL')")
    network: str = Field(description="Network identifier (e.g., 'mainnet', 'testnet')")
    name: str = Field(description="Human-readable wallet name")
    is_omnibus: bool = Field(default=False, description="Whether this is an omnibus wallet")
    comment: Optional[str] = Field(default=None, description="Optional description")
    customer_id: Optional[str] = Field(default=None, description="Optional customer ID")


class ListWalletsOptions(BaseModel):
    """Options for listing wallets."""

    limit: int = Field(default=50, ge=1, le=1000, description="Maximum items to return")
    offset: int = Field(default=0, ge=0, description="Number of items to skip")
    currency: Optional[str] = Field(default=None, description="Filter by currency")
    query: Optional[str] = Field(default=None, description="Search query")
    exclude_disabled: bool = Field(default=False, description="Exclude disabled wallets")
